### PR TITLE
Create New Splines by Extrusion and Revolution

### PR DESCRIPTION
**Pull request by jzwar**
# Creator Class
Added a `_Creator` class to be added as a member to all existing spline implementations. It provides 2 new functionalities, Spline extrusion and spline revolution.
## Extrusion
Example:
```
import gustaf as gus
spline = gus.BSpline(degrees=[1], control_points=[[0,0,0],[1,0,0]], knot_vectors=[[0,0,1,1]])
spline.create.extrude(axis=[0,0,1])
```
Some notes:
- Extraction is along `axis`
- Axis must be of same dimension as `control_points`
- extrusion is always linear, use degree elevation and knot insertion to get alternative splines

## Revolution
Example:
```
import gustaf as gus
spline = gus.BSpline(degrees=[1], control_points=[[0,0,0],[1,0,0]], knot_vectors=[[0,0,1,1]]).nurbs
spline.create.revolve(axis=[0,0,1], center=[0,0,0], angle=np.pi*0.25)
```
Some notes:
- Revolutions also work for non-rational splines, but produce a warning, as it only provides an approximation
- Beziers are only supported for up to 180° (non-including)

In summary, a way to create a cylinder would be:
```
import gustaf as gus
spline = gus.BSpline(degrees=[1], control_points=[[0,0,0],[1,0,0]], knot_vectors=[[0,0,1,1]]).nurbs
spline.create.revolve(axis=[0,0,1], center=[0,0,0], angle=np.pi*2, n_knot_spans=3).create.extrude(axis=[0,0,1])
```

### CODE REVIEW

File: gustaf/spline/create.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        if len(axis.shape) == 0:
+            axis = np.asarray([axis])
+    else:
+        raise ValueError("No extrusion Axis given")
+
+    # Check Axis dimension
+    if not (spline.control_points.shape[1] == axis.shape[0]):
+        raise ValueError(
+            "Dimension Mismatch between extrusion axis and spline."
+        )
```
Comment (j042) [inline gustaf/spline/create.py:40]:
```suggestion
    # Check Axis dimension
    # formulate correct cps
    if spline.dim == axis.shape[0]:
        cps = spline.control_points
    elif int(spline.dim + 1) == axis.shape[0]:
        # one smaller dim is allowed
        # warn that we assume new dim is all zero
        utils.log.warning(
            "Given axis is one dimension bigger than spline's dim.",
            "Assuming 0.0 entries for new dimension."
        )
        cps = np.hstack(
            (
                spline.control_points,
                np.ones((len(spline.control_points), 1)
            )
        )
    else:
        raise ValueError(
            "Dimension Mismatch between extrusion axis and spline."
        )
```

File: gustaf/spline/create.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+    # Start Extrusion
+    arguments = {}
```
Comment (j042) [inline gustaf/spline/create.py:43]:
```suggestion
    dict_spline = dict()
```
we call this kind of things `dict_spline`  + explicit `dict()`

File: gustaf/spline/create.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+
+    arguments["degrees"] = np.concatenate((spline.degrees, [1]))
+    arguments["control_points"] = np.concatenate((
+        spline.control_points,
+        spline.control_points + axis))
```
Comment (j042) [inline gustaf/spline/create.py:48]:
```suggestion
    arguments["control_points"] = np.vstack((cps, cps + axis))
```

File: gustaf/spline/create.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+Create operations for spline geometries
+"""
+
+
+import itertools
+import logging
+import numpy as np
+
+
+def extrude(spline, axis=None):
```
Comment (j042) [inline gustaf/spline/create.py:12]: Do we have better name for `axis`? with `axis`, i'd expect a magnitude separately.   For example, `along`, `extrude_vector`, ...

File: gustaf/spline/create.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    arguments["control_points"] = np.concatenate((
+        spline.control_points,
+        spline.control_points + axis))
+    if "knot_vectors" in spline._required_properties:
+        arguments["knot_vectors"] = spline.knot_vectors + [[0,0,1,1]]
+    if "weights" in spline._required_properties:
+        arguments["weights"] = np.concatenate((spline.weights, spline.weights))
+
+    return type(spline)(**arguments)
+
```
Comment (j042) [inline gustaf/spline/create.py:55]:
```suggestion


```

File: gustaf/spline/create.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+      Axis of revolution
+    center : np.ndarray
+      Center of revolution
+    angle : float
+    n_knot_spans : int
+      number of non-zero knot-elements for result-spline (if applicable)
+
+    Returns
+    -------
+    spline : spline-type
```
Comment (j042) [inline gustaf/spline/create.py:74]:
```suggestion
    spline : GustafSpline         
```
Convenient for docs

File: gustaf/spline/create.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+    # Check input type
+    if not issubclass(type(spline), GustafSpline):
+        raise NotImplementedError("Revolutions only works for splines")
+
+    # The parametric dimension is independent of the revolution but the
+    # rotation-matrix is only implemented for 2D and 3D problems
+    if not (spline.dim == 2 or spline.dim == 3):
+        raise NotImplementedError(
+            "Revolutions only implemented for 2D and 3D geometries"
```
Comment (j042) [inline gustaf/spline/create.py:86]:
```suggestion
            "Sorry,"
            "revolutions only implemented for 2D and 3D splines"
```
We gently apologize, since this is not N-dim

**Review by j042 — CHANGES_REQUESTED** [review]
very nice. thank you!  
I need a bit more time to digest rev

File: gustaf/spline/create.py (hunk 1 above)
Comment (jzwar) [reply gustaf/spline/create.py:40]: This is pretty clever, but should be zero. Also ... Why not more?

File: gustaf/spline/base.py [hunk 8]
```diff
@@ -439,6 +461,41 @@ def __init__(
 
         self._extractor = _Extractor(self)
         self._proximity = _Proximity(self)
+        self._creator = _Creator(self)
+
+    @property
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/base.py:466]: If you want to loop over a mixed list without distinguishing their types first

**Review by j042 — CHANGES_REQUESTED** [review]
NICE!
Only minor changes regarding some possible error and discussion points :)

File: tests/test_create.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+        # Test 3D revolutions for bodies
+        for spline_g in (bspline, nurbs, rationalbezier, bezier):
+            self.assertTrue(
+                np.allclose(
+                    spline_g.create.revolve(
+                        axis=[0, 0, 1],
+                        center=[0, 0, 0],
+                        angle=r_angle
+                    ).control_points[-10:, :],
```
[the remark this answers is not in the document]
Comment (jzwar) [reply tests/test_create.py:167]: Because I test the control points in the last layer of the spline. They must be rotated with respect to the original spline's control points

File: gustaf/spline/create.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+                (
+                    spline.control_points,
+                    np.zeros((len(spline.control_points), expansion_dimension))
+                )
+            )
+        else:
+            cps = np.copy(spline.control_points)
+
+        # Make sure axis is normalized
+        axis = axis / np.linalg.norm(axis)
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py:125]: Yes but it should fail. I will add an additional check to make sure that the length is non-zero

File: gustaf/spline/create.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+    else:
+        cps = np.copy(spline.control_points)
+        if spline.control_points.shape[1] == 3:
+            raise ValueError("No rotation axis given")
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py:129]: no because a rotation in 2D is always in plane -> z-axis

File: gustaf/spline/create.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+            )
+        cps -= center
+
+    # The parametric dimension is independent of the revolution but the
+    # rotation-matrix is only implemented for 2D and 3D problems
+    if not (cps.shape[1] == 2 or cps.shape[1] == 3):
+        raise NotImplementedError(
+            "Sorry,"
+            "revolutions only implemented for 2D and 3D splines"
+        )
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py:150]: Yes I did not check it before to make I get the right problem dimension which is
`max(AxisDim, CTPSdim)`

File: gustaf/spline/create.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+    else:
+        # See Rodrigues' formula
+        rotation_matrix = np.array(
+            [[0, -axis[2], axis[1]],
+             [axis[2], 0, -axis[0]],
+             [axis[1], axis[0], 0]]
+        )
+        rotation_matrix = (np.eye(3) + np.sin(rot_a) * rotation_matrix + (
+            (1 - np.cos(rot_a)) * np.matmul(rotation_matrix, rotation_matrix))
+        ).T
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py:188]: So, ... , I didn't know that this was a thing.  From your documentation I figured that they expect the angle around each cartesian axis is requested, but here I try to rotate around an arbitrary vector. Personally, I don't care which routine we use in the end, from the top of my head I just don't know how to provide the required rotation-vector...

File: gustaf/spline/create.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+
+    # Start Extrusion
+    arguments = {}
```
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py]: I forgot to push

File: gustaf/spline/create.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py]: For some inexplicable reason, division assignment does not seem to work. I have no idea why, so I just wrote:
`axis = axis / axis_norm`

File: gustaf/spline/create.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (j042) [reply gustaf/spline/create.py:194]: Ah I see, so this one is fancier.

File: gustaf/spline/create.py [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+    else:
+        # See Rodrigues' formula
+        rotation_matrix = np.array(
+            [[0, -axis[2], axis[1]],
+             [axis[2], 0, -axis[0]],
+             [axis[1], axis[0], 0]]
+        )
+        rotation_matrix = (np.eye(3) + np.sin(rot_a) * rotation_matrix + (
+            (1 - np.cos(rot_a)) * np.matmul(rotation_matrix, rotation_matrix))
+        ).T
```
Comment (j042) [inline gustaf/spline/create.py:194]:
```suggestion
        rotation_matrix = (
            np.eye(3)
            + (np.sin(rot_a) * rotation_matrix)
            + ((1 - np.cos(rot_a)) * np.matmul(rotation_matrix, rotation_matrix))
        ).T
```
btw, do you now see the origin of `eye_ders?`

File: gustaf/spline/create.py [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+    if "knot_vectors" in _RequiredProperties.of(spline):
+        spline_dict["knot_vectors"] = spline.knot_vectors + [[0, 0, 1, 1]]
+    if "weights" in _RequiredProperties.of(spline):
+        spline_dict["weights"] = np.concatenate(
+            (spline.weights, spline.weights))
+
+    return type(spline)(**spline_dict)
+
+
+def revolve(spline, axis=None, center=None, angle=None, n_knot_spans=None):
```
Comment (j042) [inline gustaf/spline/create.py:75]: The angle is expected to be radian? can we have default as degrees? It is just much friendlier

**Review by j042 — CHANGES_REQUESTED** [review]
last ones

File: gustaf/spline/create.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (j042) [reply gustaf/spline/create.py:194]: More than welcome to call this one "Next PR" if your passion level is pretty low

[the remark this answers is not in the document]
Comment (jzwar) [reply gustaf/spline/create.py:194]: No, not exactly. (utils.arr.rotate)[https://github.com/tataratat/gustaf/blob/fcf4bfdceff5ac51eb711ca6cdc55e57ac8b029f/gustaf/utils/arr.py#L252] does as `rotation_matrix`. What is called rotation axis in the script is rather the center of rotation, not the actual axis, which is given by the different angles along each axis (x,y,z). I will move this to `utils.arr` and call the script from there to make this part of the code accessible.

**Review by j042 — APPROVED** [review]
Thanks x 10^9!
I'll apply those super minor things and merge.

**Comment by jzwar** [comment]
Thanks for this last cleanup :muscle:
